### ISSUE TITLE
Terminology for Non-persistent Key vs. Persistent Key

### ISSUE THREAD

**Issue by hannestschofenig** [OpenMobileAlliance/OMA_LwM2M_for_Developers#144]
The definition for these two terms does not correspond to the common computer science definition. See

> Non-persistent Key: 
> Key that has low entropy e.g. because it is derived from a PIN or password or device serial number, or, derived from a global secret stored on multiple devices.

The definition of these two terms refers to the entropy instead (not about their persistent storage).

Now, the question is why anyone would use keys with low entropy in LWM2M altogether given that these keys are factory provisioned and/or dynamically provisioned with the bootstrapping protocol. 

For the moment I would, however, recommend to use the term low-entropy (for non-persistent) and high-entropy (for persistent) keys instead.

**Comment by hannestschofenig** [comment]
Here is proposed change: 
http://member.openmobilealliance.org/ftp/Public_documents/DM/LightweightM2M/2016/OMA-DM-LightweightM2M-2016-0130-CR_terminology_fix.zip

**Comment by boaks** [comment]
`A third question is why the entropy of the key is only related to client-initiated bootstrapping and not to server-initiated bootstrapping as well?!`

Didn't OMA-TS-LightweightM2M-V1_0-20160914-D, Page 43/44 make the same relation for that case?

By the way: If a server "initiates the bootstrap", it's unspecified, how this is mapped to DTLS roles.
Should the bootstrap server be a DTLS client? If so, the bootstrap server need to know the lwm2m endpoint of the device, otherwise, I don't see how PSK should work.
If the bootstrap server should not be the DTLS client, how does it trigger the device to open
a DTLS session and then wait for the "bootstrap"?

https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/2

https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/49

**Comment by hannestschofenig** [comment]
I am not sure which text in OMA-TS-LightweightM2M-V1_0-20160914-D, you are referring to. 

When you say "If a server "initiates the bootstrap", it's unspecified, how this is mapped to DTLS roles." what does **this** refer to? 

For the client-initiated bootstrap the LWM2M client is the DTLS client. With server-initiated bootstrapping it is true that the roles are reversed and that the LWM2M client suddenly becomes the DTLS server. Since both entities are configured with the PSK, PSK identity and also with information I don't see why DTLS with a PSK ciphersuite wouldn't work. The only challenge in that configuration is that the DTLS server requires substantially more resources (particularly RAM) than a DTLS client implementation.

**Comment by boaks** [comment]
Page 43/44

```
Step #3: Clean-up after successful Bootstrapping
The method of cleaning up the bootstrap credentials after successful bootstrapping depends on whether the LWM2M Bootstrap Server Account contains persistent or non-persistent keys. Persistent keys SHOULD be used for the LWM2M Bootstrap Server Account.
1)	LWM2M Bootstrap Server Account contains persistent keys:
In that case the original LWM2M Bootstrap Server Account SHOULD be kept.
2)	LWM2M Bootstrap Server Account contains non-persistent keys:
a)	A new LWM2M Bootstrap Server Account MUST be created during the bootstrap process - this time with a persistent keyset - for replacing the original LWM2M Bootstrap Server Account which MUST be purged by the new LWM2M bootstrap server at the end of the bootstrapping, or, only if a) cannot be applied,
b)	the LWM2M Client MUST purge the LWM2M Bootstrap Server Account after the timeout value Bootstrap Server Account Timeout, if defined, or immediately after successful bootstrapping if the timeout value is not defined.
```

I think, this is very similar to the client initiate one.

**Comment by boaks** [comment]
`Since both entities are configured with the PSK, PSK identity and also with information I don't see why DTLS with a PSK ciphersuite wouldn't work. The only challenge in that configuration is that the DTLS server requires substantially more resources (particularly RAM) than a DTLS client implementation.`

May be there are different assumptions about PSK identities.

My assumption:
The bootstrapserver has a table with lwm2m endpoint name, psk indentity, psk secret
There are many tripples in the table.
The device knows its lwm2m endpoint name, and in the security objects, also the tupples of  
psk indentity, psk secret for a host-address.

So how does the bootstrap server know, which psk identity it should use?
The device would use the host-address.

**Comment by boaks** [comment]
`When you say "If a server "initiates the bootstrap", it's unspecified, how this is mapped to DTLS roles." what does this refer to? 
`

I can't see where section 5.2.2.4 mention the DTLS role.

**Comment by dnav** [comment]
> So how does the bootstrap server know, which psk identity it should use?

In a Server-initiated bootstrap (at the beginning of the Client lifecycle), the Bootstrap Server is informed of the presence of the Client by out-of-band means (typically by the network infrastructure). So this channel can also inform the BS of the Client ID.

**Comment by hannestschofenig** [comment]
My assumption about server-initiated bootstrap is that the network somehow has to tell the bootstrap server the IP address and also has to look up what psk identity / psk it has to use with the DTLS server running on that LWM2M client.

In some sense this is very similar to how the LWM2M client acts. It has all this information provisioned upfront but in the server-initiated case the bootstrap server has to provide it dynamically since the IP address of the LWM2M client isn't know upfront.

**Comment by boaks** [comment]
Thanks!
We will see.

**Comment by hannestschofenig** [comment]
I will re-read the text of the server-initiated bootstrapping mode to see whether the text could point out the assumptions better.

**Comment by hannestschofenig** [comment]
You are right that there is some fuzziness in the text: 

"As defined in Section 5.2.3 Bootstrap Sequence, scenarios exist when the LWM2M Server is not configured within the LWM2M Client or attempts to perform the “Register” operation with LWM2M Servers have failed."

The last part of the sentence refers to this procedure: 
"
3.	If the LWM2M Client has any LWM2M Server Object Instances from the previous steps, the LWM2M Client tries to register to the LWM2M Server(s) configured in the LWM2M Server Object Instance(s).
4.	If LWM2M Client fails to register to all the LWM2M Servers or the Client doesn’t have any LWM2M Server Object Instances, and the LWM2M Client hasn’t received a Server Initiated Bootstrap within the ClientHoldOffTime, the LWM2M Client performs the Client Initiated Bootstrap. 
"

I personally think that this sequence is not practical since an IoT device is likely not going to implement multiple bootstrapping mechanisms since this will introduce a lot of complexity. In a case where the LWM2M client only supports client-initiated bootstrapping it would have to first try all these other steps first, which could take some time. Luckily, the timeout value (called ClientHoldOffTime) isn't defined in the document ;-)

**Comment by boaks** [comment]
In my opinion, the "server initiated bootstrap" as defined now, has too many "implementation specific stuff". And it make a simple and consistent implementation of "alternative path" and/or "object versioning"  "impossible".

So my favorit is (still):
* extend the client initiated bootstrap with "alternative path" and "object versioning" (together with a list of supported objects).
* define "different triggers". 
   * According to the communcation situation, the "registration failed" may be a trigger. 
   * Additionaly I would add a executable in der LWM2M server objects. 
   * And a response code 4.01 (Unauthorized) for the registration may also indicate a "credentials change" and trigger a bootstrap.
   (requires the old ones still be valid for a DTLS handshake, otherwise the client would not be able to get a reponse code)
   * There may be also "manufacturer/device" specific triggers as SMS, Button, PIN ...

**Comment by hannestschofenig** [comment]
The originally raised issue regarding non-persistent and persistent key was re-written and published in LwM2M v1.0.
